In [7]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import test
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 7: No Space Left On Device ---</h2><p>You can hear birds chirping and raindrops hitting leaves as the expedition proceeds. Occasionally, you can even hear much louder sounds in the distance; how big do the animals get out here, anyway?</p>
<p>The device the Elves gave you has problems with more than just its communication system. You try to run a system update:</p>
<pre><code>$ system-update --please --pretty-please-with-sugar-on-top
<span title="E099 PROGRAMMER IS OVERLY POLITE">Error</span>: No space left on device
</code></pre>
<p>Perhaps you can delete some files to make space for the update?</p>
<p>You browse around the filesystem to assess the situation and save the resulting terminal output (your puzzle input). For example:</p>
<pre><code>$ cd /
$ ls
dir a
14848514 b.txt
8504156 c.dat
dir d
$ cd a
$ ls
dir e
29116 f
2557 g
62596 h.lst
$ cd e
$ ls
584 i
$ cd ..
$ cd ..
$ cd d
$ ls
4060174 j
8033020 d.log
5626152 d.ext
7214296 k
</code></pre>
<p>The filesystem consists of a tree of files (plain data) and directories (which can contain other directories or files). The outermost directory is called <code>/</code>. You can navigate around the filesystem, moving into or out of directories and listing the contents of the directory you're currently in.</p>
<p>Within the terminal output, lines that begin with <code>$</code> are <em>commands you executed</em>, very much like some modern computers:</p>
<ul>
<li><code>cd</code> means <em>change directory</em>. This changes which directory is the current directory, but the specific result depends on the argument:
  <ul>
  <li><code>cd x</code> moves <em>in</em> one level: it looks in the current directory for the directory named <code>x</code> and makes it the current directory.</li>
  <li><code>cd ..</code> moves <em>out</em> one level: it finds the directory that contains the current directory, then makes that directory the current directory.</li>
  <li><code>cd /</code> switches the current directory to the outermost directory, <code>/</code>.</li>
  </ul>
</li>
<li><code>ls</code> means <em>list</em>. It prints out all of the files and directories immediately contained by the current directory:
  <ul>
  <li><code>123 abc</code> means that the current directory contains a file named <code>abc</code> with size <code>123</code>.</li>
  <li><code>dir xyz</code> means that the current directory contains a directory named <code>xyz</code>.</li>
  </ul>
</li>
</ul>
<p>Given the commands and output in the example above, you can determine that the filesystem looks visually like this:</p>
<pre><code>- / (dir)
  - a (dir)
    - e (dir)
      - i (file, size=584)
    - f (file, size=29116)
    - g (file, size=2557)
    - h.lst (file, size=62596)
  - b.txt (file, size=14848514)
  - c.dat (file, size=8504156)
  - d (dir)
    - j (file, size=4060174)
    - d.log (file, size=8033020)
    - d.ext (file, size=5626152)
    - k (file, size=7214296)
</code></pre>
<p>Here, there are four directories: <code>/</code> (the outermost directory), <code>a</code> and <code>d</code> (which are in <code>/</code>), and <code>e</code> (which is in <code>a</code>). These directories also contain files of various sizes.</p>
<p>Since the disk is full, your first step should probably be to find directories that are good candidates for deletion. To do this, you need to determine the <em>total size</em> of each directory. The total size of a directory is the sum of the sizes of the files it contains, directly or indirectly. (Directories themselves do not count as having any intrinsic size.)</p>
<p>The total sizes of the directories above can be found as follows:</p>
<ul>
<li>The total size of directory <code>e</code> is <em>584</em> because it contains a single file <code>i</code> of size 584 and no other directories.</li>
<li>The directory <code>a</code> has total size <em>94853</em> because it contains files <code>f</code> (size 29116), <code>g</code> (size 2557), and <code>h.lst</code> (size 62596), plus file <code>i</code> indirectly (<code>a</code> contains <code>e</code> which contains <code>i</code>).</li>
<li>Directory <code>d</code> has total size <em>24933642</em>.</li>
<li>As the outermost directory, <code>/</code> contains every file. Its total size is <em>48381165</em>, the sum of the size of every file.</li>
</ul>
<p>To begin, find all of the directories with a total size of <em>at most 100000</em>, then calculate the sum of their total sizes. In the example above, these directories are <code>a</code> and <code>e</code>; the sum of their total sizes is <code><em>95437</em></code> (94853 + 584). (As in this example, this process can count files more than once!)</p>
<p>Find all of the directories with a total size of at most 100000. <em>What is the sum of the total sizes of those directories?</em></p>
</article>


In [8]:
from collections import deque

from more_itertools import split_before


@dataclass
class Node:
    name: str
    size: int = field(hash=False, default=-1)

    def dirs_smaller_then(self, k, dirs: list[Directory]) -> None:
        return

    @classmethod
    def parse(cls, s: str) -> Node:
        size, name = s.split()

        if size == "dir":
            return Directory(name)

        return File(name, int(size))


@dataclass
class Directory(Node):
    children: dict[str, Node] = field(hash=False, default_factory=dict)

    def cd(self, commands: deque[list[str]]) -> None:
        self.children |= {n.split()[1]: Node.parse(n) for n in commands.popleft()[2:]}

        while commands and commands[0][0] != "$ cd ..":
            cd = commands[0][0].removeprefix("$ cd ")
            self.children[cd].cd(commands)

        if commands:
            commands.popleft()

        self.size = sum(c.size for c in self.children.values())

    def dirs_smaller_then(self, k, dirs: list[Directory]) -> None:
        if self.size <= k:
            dirs.append(self)

        for child in self.children.values():
            child.dirs_smaller_then(k, dirs)


@dataclass
class File(Node): ...


class FileSystem:
    MAX_CAPACITY = 70_000_000

    def __init__(self, s: str) -> None:
        self.root = self.parse(s)

    def parse(self, s) -> Directory:
        root = Directory("/")
        commands = deque(
            split_before(
                (l.strip() for l in s.strip().splitlines()),
                lambda l: l.startswith("$ cd "),
            )
        )
        root.cd(commands)

        return root

    def dirs_at_most(self, k=100_000) -> list[Directory]:
        dirs = []
        self.root.dirs_smaller_then(k, dirs)
        return dirs

    def total_size_dirs_at_most(self, k=100_000) -> int:
        return sum(d.size for d in self.dirs_at_most(k))

    def size_smallest_dir_freeing_enough_space(self, required: int = 30_000_000) -> int:

        return min(
            d.size
            for d in self.dirs_at_most(inf)
            if self.MAX_CAPACITY - self.root.size + d.size >= required
        )

    def __str__(self) -> str:
        return str(self.root)

In [9]:
tests = [
    {
        "name": "Example 1",
        "s": """
                $ cd /
                $ ls
                dir a
                14848514 b.txt
                8504156 c.dat
                dir d
                $ cd a
                $ ls
                dir e
                29116 f
                2557 g
                62596 h.lst
                $ cd e
                $ ls
                584 i
                $ cd ..
                $ cd ..
                $ cd d
                $ ls
                4060174 j
                8033020 d.log
                5626152 d.ext
                7214296 k
        """,
        "expected": 95437,
    },
]


@test(tests=tests)
def test_part_I(s: str) -> int:
    return FileSystem(s).total_size_dirs_at_most()


Test Example 1 passed, for test_part_I.
Success


In [10]:
with open("../input/day7.txt") as f:
    puzzle = f.read()

print(f"Part 1: {FileSystem(puzzle).total_size_dirs_at_most()}")

Part 1: 1770595


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>1770595</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>


<link href="style.css" rel="stylesheet"></link>

<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Now, you're ready to choose a directory to delete.</p>
<p>The total disk space available to the filesystem is <code><em>70000000</em></code>. To run the update, you need unused space of at least <code><em>30000000</em></code>. You need to find a directory you can delete that will <em>free up enough space</em> to run the update.</p>
<p>In the example above, the total size of the outermost directory (and thus the total amount of used space) is <code>48381165</code>; this means that the size of the <em>unused</em> space must currently be <code>21618835</code>, which isn't quite the <code>30000000</code> required by the update. Therefore, the update still requires a directory with total size of at least <code>8381165</code> to be deleted before it can run.</p>
<p>To achieve this, you have the following options:</p>
<ul>
<li>Delete directory <code>e</code>, which would increase unused space by <code>584</code>.</li>
<li>Delete directory <code>a</code>, which would increase unused space by <code>94853</code>.</li>
<li>Delete directory <code>d</code>, which would increase unused space by <code>24933642</code>.</li>
<li>Delete directory <code>/</code>, which would increase unused space by <code>48381165</code>.</li>
</ul>
<p>Directories <code>e</code> and <code>a</code> are both too small; deleting them would not free up enough space. However, directories <code>d</code> and <code>/</code> are both big enough! Between these, choose the <em>smallest</em>: <code>d</code>, increasing unused space by <code><em>24933642</em></code>.</p>
<p>Find the smallest directory that, if deleted, would free up enough space on the filesystem to run the update. <em>What is the total size of that directory?</em></p>
</article>


In [11]:
tests = [
    {
        "name": "Example 1",
        "s": """
                $ cd /
                $ ls
                dir a
                14848514 b.txt
                8504156 c.dat
                dir d
                $ cd a
                $ ls
                dir e
                29116 f
                2557 g
                62596 h.lst
                $ cd e
                $ ls
                584 i
                $ cd ..
                $ cd ..
                $ cd d
                $ ls
                4060174 j
                8033020 d.log
                5626152 d.ext
                7214296 k
        """,
        "expected": 24933642,
    },
]


@test(tests=tests)
def test_part_II(s: str) -> int:
    fs = FileSystem(s)
    return fs.size_smallest_dir_freeing_enough_space()


Test Example 1 passed, for test_part_II.
Success


<link href="style.css" rel="stylesheet"></link>


In [12]:
print(f"Part 2: {FileSystem(puzzle).size_smallest_dir_freeing_enough_space()}")

Part 2: 2195372


<link href="style.css" rel="stylesheet"></link>

<main>

<p>Your puzzle answer was <code>2195372</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>
